lib/lifecycle.js

```diff
@@ -99,82 +107,83 @@ class LifeCycle {
+            time,
+            url: pageUrl.href,
+        };
+        this.report.push(report);
 
-            // check if we have fallen off the allowed hosts
-            this.checkAllowedHost(pageConfig);
+        // get config for this page
+        let pageConfig = this.findPageConfig(pageUrl);
 
-            // collect data from this page
-            let collectedValues = await this.collectValues(pageConfig);
-            report.collect = collectedValues;
-            _.extend(this.values, collectedValues);
+        // check if we have fallen off the allowed hosts
+        this.checkAllowedHost(pageConfig);
 
-            // check if we have reached the final page
-            if (this.isFinalPage(pageConfig)) {
-                debug('FINAL PAGE');
-                return finished();
-            }
+        // collect data from this page
+        let collectedValues = await this.collectValues(pageConfig);
+        report.collect = collectedValues;
+        _.extend(this.values, collectedValues);
 
-            // check if we need to abort
-            this.checkIfExitPage(pageConfig);
-            this.checkIfMaxTriesReached(pageConfig);
+        // check if we have reached the final page
+        if (this.isFinalPage(pageConfig)) {
+            debug('FINAL PAGE');
+            return false;
+        }
 
-            // fill and submit
-            let form = new Form(this.page);
+        // check if we need to abort
+        this.checkIfExitPage(pageConfig);
+        this.checkIfMaxTriesReached(pageConfig);
+
+        // add this page visit to history
+        this.history.push(pageUrl.href);
+
+        let form = new Form(this.page);
+
+        // fill form
+        await form.fill(pageConfig.fields);
+
+        // navigate
+        await form.navigate(pageConfig.navigate);
+
+        switch (pageConfig.waitFor) {
+        case 'load':
+            debug('Waiting for page load');
+            await this.page.waitForNavigation({ timeout: pageConfig.navigateTimeout, waitUntil: 'load' });
+            break;
+        case 'idle':
+            debug('Waiting for network idle');
+            await this.page.waitForNavigation({ timeout: pageConfig.navigateTimeout, waitUntil: 'networkidle2' });
+            break;
+        default:
+            debug(`Waiting ${pageConfig.waitFor}ms`);
+            await this.delay(pageConfig.waitFor);
+        }
 
-            if (pageConfig.slowMo) await this.delay(pageConfig.slowMo / 2);
-            await form.fill(pageConfig.fields);
+        return true;
+    }
 
-            if (pageConfig.slowMo) await this.delay(pageConfig.slowMo / 2);
-            await form.submit(pageConfig.submit, pageConfig.submitTimeout);
+    async run() {
+        debug('Run', this.config.start.href);
 
-            // add this page visit to history
-            this.history.push(pageUrl.href);
+        try {
+            await this.page.goto(this.config.start.href);
+            while (await this.processPage());
```

Comment:
move back to awaiting for page processing to complete instead of hanging off the load event